Building quality apps/Unit tests.md

```diff
@@ -0,0 +1,155 @@
+    // Act
+    val actualUser = loginUseCase.login(username, password)
+    // Assert
+    assertEquals(expectedUser, actualUser)
+}
+```
+
+The whole test is so much nicer now, don't you think? 
+As a side note, when writing tests you do not need to write the actual section names, it is ok to just leave an empty space between those sections.
+
+## Writing good unit tests
+
+Writing good tests is sometimes hard and sometimes very hard. Nevertheless, we will try to define some general guidelines to help us avoid common mistakes and achieve better tests at the end of the day.
+
+### Use test doubles
+
+As mentioned in the previous section we often have to arrange some behaviours or objects that we usually have in production code, which can sometimes be tricky depending on the complexity of a test. **In order to replace some production code for testing purposes we use a generic term called Test Double.**
+There are several test doubles that are worth mentioning and which you will use the most and those are:
+
+* Mock - An object which we use for verification by checking whether particular methods were called, with specific parameters.
+* Stub - An object that holds predefined data and uses it to answer calls during tests. It is used when we cannot or don’t want to involve objects that would answer with real data or have undesirable side effects.
+* Fake - An object that has a working implementation, but not same as production one. Usually it takes some shortcut and has a simplified version of the production code. For example faking a database with an in memory solution in order to make our unit tests faster.
+
+
+When we talk about mocks and stubs we do not need to write our own implementation of those test doubles. Luckily, we have a tool called Mockito which enables us to mock and stub behaviours and dependencies with ease. Keep in mind that Mockito does not enable us to create fakes.
+
+Let's take a look at two examples where we use Mockito stubbing and mocking features.
+
+* Using Mockito for mocking
+
+```kotlin  
+val authTokenManager = mock(AuthTokenManager::class.java)
+verify(authTokenManger).saveAuthToken("Token")
+```
+With mocks we are verifying a specific behavior which can be seen in the above example using the verify method from Mockito.
+
+* Using Mockito for stubbing
+
+```kotlin  
+val authTokenManager = mock(AuthTokenManager::class.java)
+whenever(authTokenManager.getAuthToken()).thenReturn("Token")
+```
+With stubs we predefine results for a specific method so that just like it is shown in the above example. Every time the `getAuthToken` method is called we will get a stubbed result which is in our case `"Token"`.
+
+
+It is important to know this terminology because it is essentially part of the  testing fundamentals and it will definitely help you on your journey in becoming better in writing tests. You will also often find these terms in documentation of some testing library that you will use.
```

Comment:
```suggestion
It is important to know this terminology because it is essentially part of the  testing fundamentals and it will definitely help you on your journey to becoming better at writing tests. You will also often find these terms in documentation of some testing library that you will use.
```